File: app/src/main/java/io/github/davoleo/barcodestock/scanner/BarcodeScannerActivity.java
package io.github.davoleo.barcodestock.scanner;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.*;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import android.util.Size;
import android.view.Surface;
import android.view.TextureView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import com.google.android.gms.tasks.Task;
import com.google.mlkit.vision.barcode.Barcode;
import com.google.mlkit.vision.barcode.BarcodeScanner;
import com.google.mlkit.vision.barcode.BarcodeScannerOptions;
import com.google.mlkit.vision.barcode.BarcodeScanning;
import io.github.davoleo.barcodestock.R;
import io.github.davoleo.barcodestock.ui.MainActivity;
import io.github.davoleo.barcodestock.util.CompareSizeByArea;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BarcodeScannerActivity extends AppCompatActivity {

    private static final int CAMERA_REQUEST_PERMISSION_RESULT = 0;

    private BarcodeScannerOptions formats = new BarcodeScannerOptions.Builder()
            .setBarcodeFormats(Barcode.FORMAT_EAN_13, Barcode.FORMAT_UPC_A, Barcode.FORMAT_UPC_E)
            .build();

    private String cameraId;
    private Size previewSize;
    //Represents a mobile camera device
    private CameraDevice cameraDevice;
    //Return a specific Camera Device
    private CameraDevice.StateCallback cameraDeviceStateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(@NonNull CameraDevice camera) {
            cameraDevice = camera;
            startPreview();
        }

        //Clean up memory on disconnect and error
        @Override
        public void onDisconnected(@NonNull CameraDevice camera) {
            camera.close();
            cameraDevice = null;
        }

        @Override
        public void onError(@NonNull CameraDevice camera, int error) {
            camera.close();
            cameraDevice = null;
        }
    };

    private HandlerThread backgroundThread;
    private Handler backgroundHandler;

    private CaptureRequest.Builder captureRequestBuilder;

    //Components
    private TextureView textureView;
    //Notifies when the textureView is available
    private TextureView.SurfaceTextureListener surfaceTextureListener = new TextureView.SurfaceTextureListener() {
        @Override
        public void onSurfaceTextureAvailable(@NonNull SurfaceTexture surfaceTexture, int width, int height) {
            setupCamera(width, height);
            connectCamera();
        }

        @Override
        public void onSurfaceTextureSizeChanged(@NonNull SurfaceTexture surfaceTexture, int width, int height) {
        }

        @Override
        public boolean onSurfaceTextureDestroyed(@NonNull SurfaceTexture surfaceTexture) {
            return true;
        }

        @Override
        public void onSurfaceTextureUpdated(@NonNull SurfaceTexture surfaceTexture) {
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_barcode_scanner);

        textureView = findViewById(R.id.cameraPreview);
    }

    @Override
    protected void onResume() {
        super.onResume();

        startBackgroundThread();

        //Checks if the textureview is available
        if (textureView.isAvailable()) {
            setupCamera(textureView.getWidth(), textureView.getHeight());
            connectCamera();
        }
        else {
            textureView.setSurfaceTextureListener(surfaceTextureListener);
        }
    }

    @Override
    protected void onPause() {
        closeCamera();
        stopBackgroundThread();
        super.onPause();
    }

    private void setupCamera(int width, int height) {
        CameraManager manager = ((CameraManager) getSystemService(Context.CAMERA_SERVICE));

        assert manager != null;
        //Loop through all the available camera IDs (usually front and rear but there could be more on some devices)
        try {
            for (String id : manager.getCameraIdList()) {
                CameraCharacteristics characteristics = manager.getCameraCharacteristics(id);

                if (characteristics.get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_FRONT)
                    continue;

                StreamConfigurationMap configMap = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);

                int portraitWidth = height;
                int portraitHeight = width;

                //Swap the values because the app is always in portrait mode (and the width and height are calculated in landscape mode)
                previewSize = chooseOptimalSize(configMap.getOutputSizes(SurfaceTexture.class), width, height);
                cameraId = id;
                return;
            }
        } catch (CameraAccessException e) {
            Log.e(MainActivity.TAG, "setupCamera: Something Went wrong while accessing the Camera Devices", e);
        }
    }

    private void connectCamera() {
        CameraManager cameraManager = ((CameraManager) getSystemService(Context.CAMERA_SERVICE));

        try {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                cameraManager.openCamera(cameraId, cameraDeviceStateCallback, backgroundHandler);
            } else {
                if (shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
                    Toast.makeText(this, "Barcode Scanner requires access to camera", Toast.LENGTH_SHORT).show();
                }
                requestPermissions(new String[] {Manifest.permission.CAMERA}, CAMERA_REQUEST_PERMISSION_RESULT);
            }
        } catch (CameraAccessException e) {
            Log.e(MainActivity.TAG, "connectCamera: There was a problem connecting to the Camera", e);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == CAMERA_REQUEST_PERMISSION_RESULT) {
            if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Barcode Scanner requires access to camera", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void startPreview() {
        SurfaceTexture surfaceTexture = textureView.getSurfaceTexture();
        surfaceTexture.setDefaultBufferSize(previewSize.getWidth(), previewSize.getHeight());
        Surface previewSurface = new Surface(surfaceTexture);

        try {
            captureRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            captureRequestBuilder.addTarget(previewSurface);

            cameraDevice.createCaptureSession(Collections.singletonList(previewSurface), new CameraCaptureSession.StateCallback() {
                @Override
                public void onConfigured(@NonNull CameraCaptureSession session) {
                    try {
                        session.setRepeatingRequest(captureRequestBuilder.build(), null, backgroundHandler);
                    } catch (CameraAccessException e) {
                        Log.e(MainActivity.TAG, "startPreview: There was a problem accessing the camera", e);
                    }
                }

                @Override
                public void onConfigureFailed(@NonNull CameraCaptureSession session) {
                    Toast.makeText(BarcodeScannerActivity.this, "Unable to setup camera preview correctly", Toast.LENGTH_SHORT).show();
                }
            }, null);
        } catch (CameraAccessException e) {
            Log.e(MainActivity.TAG, "startPreview: There was a problem accessing the camera", e);
        }
    }

    private void closeCamera() {
        if (cameraDevice != null) {
            cameraDevice.close();
            cameraDevice = null;
        }
    }

    //Setup the background thread (to avoid doing things on the UI Thread)
    private void startBackgroundThread() {
        backgroundThread = new HandlerThread("BarcodeStockCameraService");
        backgroundThread.start();
        backgroundHandler = new Handler(backgroundThread.getLooper());
    }

    private void stopBackgroundThread() {
        backgroundThread.quitSafely();
        try {
            backgroundThread.join();
        } catch (InterruptedException e) {
            Log.e(MainActivity.TAG, "stopBackgroundThread: The thread was interrupted while waiting to die :c", e);
            e.printStackTrace();
        } finally {
            backgroundThread = null;
            backgroundHandler = null;
        }
    }

    private static Size chooseOptimalSize(Size[] choices, int width, int height) {
        List<Size> bigEnough = new ArrayList<>();

        for (Size size : choices) {
            //check if the ratio matches and that the supported size is greater than the preview size
            if (/*size.getHeight() == size.getWidth() * height / width &&*/ size.getWidth() >= width && size.getHeight() >= height) {
                bigEnough.add(size);
            }
        }

        if (!bigEnough.isEmpty()) {
            return Collections.min(bigEnough, new CompareSizeByArea());
        } else {
            Log.w(MainActivity.TAG, "chooseOptimalSize: NO COMPATIBLE SIZE", null);
            return choices[0];
        }
    }

    private void scanBarcode() {
        BarcodeScanner scanner = BarcodeScanning.getClient(formats);

        Task<List<Barcode>> result = scanner.process(null);
    }

}

File: app/src/main/java/io/github/davoleo/barcodestock/scanner/BarcodeScannerActivity.kt
package io.github.davoleo.barcodestock.scanner

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Rect
import android.graphics.RectF
import android.media.RingtoneManager
import android.os.Bundle
import android.util.Log
import android.util.Size
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.camera2.Camera2Config
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.mlkit.vision.barcode.Barcode
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import io.github.davoleo.barcodestock.R
import io.github.davoleo.barcodestock.ui.MainActivity
import kotlinx.android.synthetic.main.activity_barcode_scanner.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.max
import kotlin.math.min

class BarcodeScannerActivity : AppCompatActivity(), ImageAnalysis.Analyzer {

    private val formats = BarcodeScannerOptions.Builder()
            .setBarcodeFormats(Barcode.FORMAT_EAN_13, Barcode.FORMAT_UPC_A, Barcode.FORMAT_UPC_E)
            .build()

    private var barcodeScanner: BarcodeScanner = BarcodeScanning.getClient(formats)

    private var preview: Preview? = null
    private var graphicOverlay: GraphicOverlay? = null
    private var imageAnalyzer: ImageAnalysis? = null
    private var camera: Camera? = null

    private val targetResolution = Size(480, 640)

    private var delayCount = 0

    private lateinit var cameraExecutor: ExecutorService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_barcode_scanner)

        lifecycle.addObserver(barcodeScanner)
        CameraX.configureInstance(Camera2Config.defaultConfig())

        setSupportActionBar(findViewById(R.id.toolbar))

        //Request Camera Permissions
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED)
        {
            startCamera()
        }
        else
        {
            if (shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
                Toast.makeText(this, "Barcode Scanner requires access to the camera", Toast.LENGTH_SHORT).show()
            }
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), CAMERA_PERMISSION_REQUEST_CODE)
        }

        CameraX.getOrCreateInstance(applicationContext)

        graphicOverlay = findViewById<GraphicOverlay>(R.id.cameraPreviewOverlay).apply {
            setOnClickListener { //overlay: View? ->
                run {
                    Log.i(MainActivity.TAG, "you tapped the OVERLAY :o")
                }
            }
        }

        cameraExecutor = Executors.newSingleThreadExecutor()

        imageAnalyzer = ImageAnalysis.Builder()
                .setTargetResolution(targetResolution)
                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                .build()
                .also {
                    it.setAnalyzer(cameraExecutor, this)
                }
    }

    override fun onDestroy() {
        cameraExecutor.shutdown()
        CameraX.shutdown()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (requestCode == CAMERA_PERMISSION_REQUEST_CODE) {
            if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Barcode Scanner requires access to the camera", Toast.LENGTH_SHORT).show()
                finish()
            } else {
                startCamera()
            }
        }
    }

    private fun startCamera() {
        val futureCameraProvider = ProcessCameraProvider.getInstance(this)

        futureCameraProvider.addListener(Runnable {
            //Used to bind lifecycle of cameras to the lifecycle owner [takes care of opening and closing the camera]
            val cameraProvider: ProcessCameraProvider = futureCameraProvider.get()

            //preview
            preview = Preview.Builder().build()

            val cameraSelector = CameraSelector.Builder().requireLensFacing(CameraSelector.LENS_FACING_BACK).build()

            try {
                //Unbind all previous use cases before rebinding
                cameraProvider.unbindAll()

                //Bind our use cases to camera
                camera = cameraProvider.bindToLifecycle(this, cameraSelector, preview, imageAnalyzer)
                preview?.setSurfaceProvider(cameraPreview.createSurfaceProvider())
            } catch (e: Exception) {
                Log.e(MainActivity.TAG, "There was an issue binding the new use case", e)
            }

        }, ContextCompat.getMainExecutor(this))
    }

    @ExperimentalGetImage
    override fun analyze(imageProxy: ImageProxy) {

        val mediaImage = imageProxy.image

        if (mediaImage != null) {
            val image = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)

            if (graphicOverlay != null) {
                barcodeScanner.process(image)
                        .addOnSuccessListener { barcodeList ->
                            onSuccess(barcodeList, graphicOverlay!!)
                        }
                        .addOnFailureListener {
                            onFailure(it)
                        }
                        .addOnCompleteListener { imageProxy.close() }
            }
            else
                Log.w(MainActivity.TAG, "overlay is null :(")
        }
    }

    fun onSuccess(results: List<Barcode>, graphicOverlay: GraphicOverlay) {

        //Debug Logging ---
        //Log.i(MainActivity.TAG, "Barcode Type: " + results.firstOrNull()?.valueType)

        val reticle = BarcodeGraphic.getReticleBox(graphicOverlay)
        val widthScaleFactor: Float = graphicOverlay.width / targetResolution.width.toFloat()
        val heightScaleFactor: Float = graphicOverlay.height / targetResolution.height.toFloat()
        val barcodeInCenter: Barcode?

        barcodeInCenter = results.firstOrNull { barcode ->
            //Take barcode's bounding box, otherwise if boundingBox is null then return null as barcodeInCenter
            val boundingBox = barcode.boundingBox ?: return@firstOrNull false
            val box = translateRect(boundingBox, widthScaleFactor, heightScaleFactor)
            box.contains(reticle.centerX(), reticle.centerY())
        }

        //Debug Logging ---
        //Log.i(MainActivity.TAG, "onSuccess: $barcodeInCenter")
        //Log.i(MainActivity.TAG, "Raw Value: $barcodeInCenter")
        //val xLeft = results.firstOrNull()?.boundingBox?.left ?: -1
        //val yTop = results.firstOrNull()?.boundingBox?.top ?: -1
        //val xRight = results.firstOrNull()?.boundingBox?.right ?: -1
        //val yBottom = results.firstOrNull()?.boundingBox?.bottom ?: -1
        //Log.i(MainActivity.TAG, "barcode LTRB: $xLeft, $yTop, $xRight, $yBottom")
        //Log.i(MainActivity.TAG, "center XY: ${reticle.centerX()}, ${reticle.centerY()}")
        //Log.i(MainActivity.TAG, "center XY: ${reticle.width() / 2}, ${reticle.height() / 2}")
        graphicOverlay.clear()

        if (barcodeInCenter == null) {
            delayCount = max(0, --delayCount)
            graphicOverlay.add(BarcodeGraphic(graphicOverlay, ContextCompat.getColor(applicationContext, R.color.colorPrimaryLight)))
        }
        else {
            graphicOverlay.add(BarcodeGraphic(graphicOverlay, ContextCompat.getColor(applicationContext, R.color.pureRed)))
            delayCount = min(2, ++delayCount)

            if (delayCount == 2) {
                delayCount = 0
                val notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
                val notifySound = RingtoneManager.getRingtone(applicationContext, notification)
                notifySound.play()

                val intent = Intent()
                intent.putExtra("barcode", barcodeInCenter.rawValue)
                setResult(Activity.RESULT_OK, intent)
                finish()
            }
        }

        graphicOverlay.invalidate()
    }

    fun onFailure(e: Exception) {
        Log.e(MainActivity.TAG, "Barcode detection failed!", e)
    }

    companion object {
        private const val CAMERA_PERMISSION_REQUEST_CODE = 0

        /**
         * It doubles the size of the rectangle keeping the ratio if the multiplier is 1
         */
        fun increaseSize(rect: Rect, multiplier: Float): RectF {
            val right = rect.right + (multiplier * rect.width())
            val bottom = rect.bottom + (multiplier * rect.height())
            return RectF(rect.left.toFloat(), rect.top.toFloat(), right, bottom)
        }

        fun translateRect(rect: Rect, widthFactor: Float, heightFactor: Float): RectF
        {
            val left = rect.left * widthFactor
            val top = rect.top * heightFactor
            val right = rect.right * widthFactor
            val bottom = rect.bottom * heightFactor

            //Debug Prints ---
            //Log.i(MainActivity.TAG, "translateRect: Factors: $widthFactor, $heightFactor")
            Log.i(MainActivity.TAG, "translateRect LTRB: $left, $top, $right, $bottom")

            return RectF(left, top, right, bottom)
        }
    }
}
